### PR TITLE
HHH-9731 : Potential bug in SearchedCaseExpression type handling

### PR DESCRIPTION
**Pull request by dreab8**
https://hibernate.atlassian.net/browse/HHH-9731

### CODE REVIEW

File: hibernate-entitymanager/src/main/java/org/hibernate/jpa/criteria/expression/SearchedCaseExpression.java [hunk 1]
```diff
@@ -74,13 +75,16 @@ public SearchedCaseExpression(
 	}
 
 	public Case<R> when(Expression<Boolean> condition, Expression<? extends R> result) {
-		WhenClause whenClause = new WhenClause( condition, result );
+		WhenClause whenClause = new WhenClause(
+				condition,
+				new CastFunction( criteriaBuilder(), result.getJavaType(), (ExpressionImpl) result )
```
Comment (sebersole) [inline hibernate-entitymanager/src/main/java/org/hibernate/jpa/criteria/expression/SearchedCaseExpression.java]: Let's add some comments here about why we are wrapping this in a cast.  It is needed for the HQL parser, but that is not apparent/obvious.

File: hibernate-entitymanager/src/main/java/org/hibernate/jpa/criteria/expression/SearchedCaseExpression.java [hunk 2]
```diff
@@ -88,11 +92,11 @@ private void adjustJavaType(Expression<? extends R> exp) {
 	}
 
 	public Expression<R> otherwise(R result) {
-		return otherwise( buildLiteral(result) );
+		return otherwise( buildLiteral( result ) );
 	}
 
 	public Expression<R> otherwise(Expression<? extends R> result) {
-		this.otherwiseResult = result;
+		this.otherwiseResult = new CastFunction( criteriaBuilder(), result.getJavaType(), (ExpressionImpl) result );
```
Comment (sebersole) [inline hibernate-entitymanager/src/main/java/org/hibernate/jpa/criteria/expression/SearchedCaseExpression.java]: Same as above

File: hibernate-entitymanager/src/main/java/org/hibernate/jpa/criteria/expression/SimpleCaseExpression.java [hunk 3]
```diff
@@ -84,7 +85,7 @@ public SimpleCaseExpression(
 	public SimpleCase<C, R> when(C condition, Expression<? extends R> result) {
 		WhenClause whenClause = new WhenClause(
 				new LiteralExpression<C>( criteriaBuilder(), condition ),
-				result
+				new CastFunction( criteriaBuilder(), result.getJavaType(), (ExpressionImpl) result )
```
Comment (sebersole) [inline hibernate-entitymanager/src/main/java/org/hibernate/jpa/criteria/expression/SimpleCaseExpression.java]: And again, and below :)

**Comment by sebersole** [comment]
Overall I think it would be best to only do the casts in the rendered HQL.  Let's look at this together after I get 5.0.1 done.

**Comment by dreab8** [comment]
@sebersole 
https://github.com/hibernate/hibernate-orm/pull/1072 a different solution for this issue

**Comment by dreab8** [comment]
applied https://github.com/hibernate/hibernate-orm/pull/1072

**Pull request closed without merging**